api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java

```diff
@@ -211,30 +213,92 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
   }
 
-  public boolean isCompleteAndNotExpired(Timestamp completionTime) {
-    Timestamp expirationTime =
-        new Timestamp(
-            clock.millis()
-                - TimeUnit.MILLISECONDS.convert(
-                    configProvider.get().accessRenewal.expiryDays, TimeUnit.DAYS));
-    if (configProvider.get().access.enableAccessRenewal) {
-      return completionTime != null && expirationTime.before(completionTime);
+  private class ModuleTimes {
+    public Optional<Timestamp> completion;
+    public Optional<Timestamp> bypass;
+
+    public ModuleTimes(Timestamp nullableCompletion, Timestamp nullableBypass) {
+      completion = Optional.ofNullable(nullableCompletion);
+      bypass = Optional.ofNullable(nullableBypass);
+    }
+
+    public boolean isBypassed() {
+      return bypass.isPresent();
+    }
+
+    public boolean isComplete() {
+      return completion.isPresent();
+    }
+
+    public Optional<Timestamp> getExpiration() {
+      if (isBypassed()) {
+        return Optional.empty();
+      }
+      if (!completion.isPresent()) {
+        return Optional.empty();
+      }
+      if (!configProvider.get().access.enableAccessRenewal) {
+        return Optional.empty();
+      }
+      Long expiryDays = configProvider.get().accessRenewal.expiryDays;
+      if (expiryDays == null) {
+        throw new RuntimeException("config value accessRenewal.expiryDays.expiryDays is null");
+      }
+      long expiryDaysInMs = TimeUnit.MILLISECONDS.convert(expiryDays, TimeUnit.DAYS);
+      return Optional.of(new Timestamp(completion.get().getTime() + expiryDaysInMs));
+    }
+
+    public boolean hasExpired() {
+      final Timestamp now = new Timestamp(clock.millis());
+      return getExpiration().map(x -> x.before(now)).orElse(false);
+    }
+
+    public boolean areCompliant() {
+      return isBypassed() || (isComplete() && !hasExpired());
     }
-    return completionTime != null;
+  }
+
+  private RenewableAccessModuleStatus mkStatus(ModuleNameEnum name, ModuleTimes times) {
+    return new RenewableAccessModuleStatus()
+        .moduleName(name)
+        .expirationEpochMillis(times.getExpiration().map(x -> x.getTime()).orElse(null))
+        .hasExpired(times.hasExpired());
+  }
+
+  public List<RenewableAccessModuleStatus> getRenewableAccessModuleStatus(DbUser user) {
+    return ImmutableList.of(
+        mkStatus(
+            ModuleNameEnum.COMPLIANCETRAINING,
+            new ModuleTimes(
+                user.getComplianceTrainingCompletionTime(),
+                user.getComplianceTrainingBypassTime())),
+        mkStatus(
+            ModuleNameEnum.DATAUSEAGREEMENT,
+            new ModuleTimes(
+                user.getDataUseAgreementCompletionTime(), user.getDataUseAgreementBypassTime())),
+        mkStatus(
+            ModuleNameEnum.PROFILECONFIRMATION,
+            new ModuleTimes(user.getProfileLastConfirmedTime(), null)),
```

Comment:
alt: single param constructor?  (I don't know what my preference is)

Reply:
Good thought. Also split.